### PR TITLE
Add Kotlin compiler formula.

### PR DESCRIPTION
**Pull request by pardom**

### CODE REVIEW

File: Library/Formula/kotlin-compiler.rb [hunk 1]
```diff
@@ -0,0 +1,17 @@
+class KotlinCompiler < Formula
+  homepage "http://kotlinlang.org/"
+  url "https://github.com/JetBrains/kotlin/releases/download/M11.1-bootstrap/kotlin-compiler-0.11.91.1.zip"
+  sha256 "5a68aa0f42d1e14f28fdcbf317d873fdbac1ef736934f7877fa26dac64377f64"
+
+
```
Comment (DomT4) [inline Library/Formula/kotlin-compiler.rb]: Only one blank line necessary :)

File: Library/Formula/kotlin-compiler.rb [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+  def install
+    libexec.install %w[bin lib]
+    bin.install_symlink "#{libexec}/bin/kotlinc"
```
Comment (DomT4) [inline Library/Formula/kotlin-compiler.rb]: If there's nothing else in the ` bin ` you can just do `bin.install_symlink Dir["#{libexec}/bin/*"]`

File: Library/Formula/kotlin-compiler.rb [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+    bin.install_symlink "#{libexec}/bin/kotlinc-js"
+    bin.install_symlink "#{libexec}/bin/kotlinc-jvm"
+  end
+
+  test do
+    system "#{bin}/kotlinc", "-version"
```
Comment (DomT4) [inline Library/Formula/kotlin-compiler.rb]: Can this test be modified to do something more substantial than e.g. --version or --help? See `cmake.rb` for an example of an application formula with a good test and `tinyxml2.rb` for an example of a library formula with a good test. Thanks!

Comment (pardom) [inline Library/Formula/kotlin-compiler.rb]: Sure. I saw this in another formula and did the same. Is compilation and output assertion of a Kotlin file sufficient?

Comment (DomT4) [inline Library/Formula/kotlin-compiler.rb]: Yeah, that'd be awesome :tada:.

File: Library/Formula/kotlin-compiler.rb (hunk 2 above)
Comment (pardom) [inline Library/Formula/kotlin-compiler.rb]: There are `.bat` files in the directory. The following succeeds for audit, but not install:

```ruby
bin.install_symlink Dir["#{libexec}/bin/*"].exclude("*.bat")
```

Comment (pardom) [inline Library/Formula/kotlin-compiler.rb]: Never mind, this works:

```ruby
def install
    rm Dir['*.bat']
    libexec.install %w[bin lib]
    bin.install_symlink Dir["#{libexec}/bin/*"]
  end
```

Comment (DomT4) [inline Library/Formula/kotlin-compiler.rb]: Oh you can get rid of them first.

` rm_rf Dir['*.bat'] ` or similar.

Comment (DomT4) [inline Library/Formula/kotlin-compiler.rb]: > Never mind, this works:

:+1:

Comment (DomT4) [inline Library/Formula/kotlin-compiler.rb]: Make sure you double quote `"*.bat"` though :wink:. Single quotes around anything are an old thing.

**Comment by pardom** [comment]
Seems like a strange failure, and only on Mountain Lion.

`fatal: unable to access 'https://github.com/Homebrew/homebrew.git/': Could not resolve host: (nil); nodename nor servname provided, or not known`

http://bot.brew.sh/job/Homebrew%20Pull%20Requests/25211/version=mountain_lion/console

**Comment by tdsmith** [comment]
Thanks for the pull request! 🎉 Homebrew depends on contributions from community members like you and we're grateful for your support.

**Comment by udalov** [comment]
@pardom why not just `kotlin`?

**Comment by pardom** [comment]
Good question. I didn't change the name `brew create` gave it based on the url.

**Comment by DomT4** [comment]
Yeah, we can't change formula names at this point - It breaks everything, horribly so. You could always tee up a PR to add an `alias` if desired, so people doing `brew install kotlin` end up with the expected formula. No guarantee it'll be accepted, but only one way to find out :wink:.

**Comment by pardom** [comment]
I'll make sure to link your comment in the PR. :sweat_smile:

**Comment by udalov** [comment]
@DomT4 What do you mean it breaks everything? Wouldn't just deleting the `kotlin-compiler` formula and adding a similar `kotlin` formula work? As far as I understand, formulae are added/deleted all the time.

**Comment by DomT4** [comment]
There's a lot of issues on this that go into detail if you search for terms like "formula rename" throughout the Github issues, but #14374 is the origin. Renames have caused *major* issues before. It's a known bug.

Deleting and readding is slightly different, but comes with a new set of problems - Nobody who currently has it installed will be notified of that happening, and 99% won't notice and will never, ever get updates for the formula again.

The new formula would also conflict with the old formula, i.e. it'd obviously link all the same names into the same common prefix, but we can't add a `conflicts_with` for a formula that no longer exists, so anyone who had previously done `brew install kotlin-compiler` but hadn't undone that prior to doing `brew install kotlin` would end up with some very nasty looking linking errors which require user intervention to resolve.

There's a very solid chance that people who don't read the update log, or don't watch the Homebrew repo, would never notice the change and that creates problems.

**Comment by udalov** [comment]
Thanks for a detailed explanation!

I wonder if deleting/adding is in fact a big deal in this particular case: Kotlin web site was silently updated only 12 hours ago (as of now) to include instructions on the Homebrew installation: https://github.com/JetBrains/kotlin-web-site/pull/192. Judging by the number of our users from the other sources, I'm guessing not a lot of people had a chance to find out about this formula, or even install it. However I'm not counting those Homebrew users who just installed it out of curiousity. Still I think it's not too late to delete/add the formula. Or are such dangerous actions completely discouraged in Homebrew?

**Comment by DomT4** [comment]
Anything that breaks the user experience is usually frowned upon. Honestly, an alias is going to do most of what you desired as a resolution here - You'll be able to do either `brew install kotlin` or `brew install kotlin-compiler` and end up with the same result. That's been the standard way to "rename" formulae without actually being able to rename them for a while. If you feel strongly about it you can ping one of the maintainers, Jack probably, but I can't see them being up for it given the risks.

**Comment by udalov** [comment]
Got it, thanks. I guess an alias is OK indeed.

**Pull request closed without merging**